invoke : ensure custom envs of CNIArgs are prepended to process envs

mars1024 opened the pull request:
From this commit [https://github.com/golang/go/commit/e73f4894949c4ced611881329ff8f37805152585#diff-2f929a7660c606fd0c5b268dad1a9bd4R61](https://github.com/golang/go/commit/e73f4894949c4ced611881329ff8f37805152585#diff-2f929a7660c606fd0c5b268dad1a9bd4R61) in golang, we can know that the RawExec will clean the duplicated envs and keep the last valid env of the same key.
So if we want to prepend custom args from CNIArgs rather than the existing process environments, we should append custom args in the end.

mars1024 commented:
another ref [https://github.com/golang/go/issues/12868](https://github.com/golang/go/issues/12868) , I think some code changes in golang Exec lead to this bug.

mars1024 commented:
I think it's the original issue [https://github.com/containernetworking/cni/issues/410](https://github.com/containernetworking/cni/issues/410) related to this, if you have time, please take a look at this PR, thanks ~ @bboreham @rosenhouse

bboreham commented:
Thanks for the PR!
#410 was fixed by #461, but noted near the time to be broken again by a Go runtime change.
We should have opened another issue.
Wonder if you can change the test so it actually tests the end result?

mars1024 commented:
> Thanks for the PR!
> #410 was fixed by #461, but noted near the time to be broken again by a Go runtime change.
> We should have opened another issue.
> Wonder if you can change the test so it actually tests the end result?

I don't think we can test the end result, because the deduplicate action happens in exec or delegate, not here, but I think there maybe another way, we can add a deduplicate action in func AsEnv to make sure there is no duplication in return.

mars1024 commented:
I have add deduplication in envs return of CNIArgs and updated the tests. @bboreham

dcbw commented:
@mars1024 Instead of using an array in AsEnv, what about just decomposing the environment into a map, adding our CNI_ vars into the map, and recomposing it into the array we finally pass back?

```
envMap := make(map[string]string)
env := os.GetEnv()
pairs := strings.Split(args, ";")
for _, pair := range pairs {
	kv := strings.Split(pair, "=")
	if len(kv) != 2 || kv[0] == "" || kv[1] == "" {
		return nil, fmt.Errorf("invalid CNI_ARGS pair %q", pair)
	}
	envMap[kv[0]] = kv[1]
}

envMap["CNI_COMMAND"] = <xxxx>
envMap["CNI_CONTAINERID"] = <xxx>
```
I don't think we need to bother with case sensitivity because CNI defines the relevant enviornment variables that we are adding here as upper-case. If something bothers to look for "cni_command" then that thing is demonstrably wrong and not spec conformant. (unless something on Windows does env lookups case insensitively and grabs the first item in the list?)

mars1024 commented:
> @mars1024 Instead of using an array in AsEnv, what about just decomposing the environment into a map, adding our CNI_ vars into the map, and recomposing it into the array we finally pass back?
> 
> ```
> envMap := make(map[string]string)
> env := os.GetEnv()
> pairs := strings.Split(args, ";")
> for _, pair := range pairs {
> 	kv := strings.Split(pair, "=")
> 	if len(kv) != 2 || kv[0] == "" || kv[1] == "" {
> 		return nil, fmt.Errorf("invalid CNI_ARGS pair %q", pair)
> 	}
> 	envMap[kv[0]] = kv[1]
> }
> 
> envMap["CNI_COMMAND"] = <xxxx>
> envMap["CNI_CONTAINERID"] = <xxx>
> 
> finalEnv := make([]string, 0, len(envMap))
> for k, v := range envMap {
>     finalEnv = append(finalEnv, fmt.Sprintf("%s=%s", k, v))
> }
> return finalEnv
> ```
> 
> I don't think we need to bother with case sensitivity because CNI defines the relevant enviornment variables that we are adding here as upper-case. If something bothers to look for "cni_command" then that thing is demonstrably wrong and not spec conformant. (unless something on Windows does env lookups case insensitively and grabs the first item in the list?)

OK, thanks for your comment, I will update the deduplication func.

mars1024 commented:
Code updated! @bboreham @dcbw @squeed  Please go on reviewing, thanks.

dcbw commented:
/lgtm
